src/cli/parse-flags.ts

```diff
@@ -54,132 +51,254 @@ export const parseFlags = (args: string[], _sys?: CompilerSystem): ConfigFlags =
-
-  if (typeof flags[configCaseName] !== 'boolean') {
-    flags[configCaseName] = null;
+const parseCLITerm = (flags: ConfigFlags, args: string[]) => {
+  // pull off the first arg from the argument array
+  const arg = args.shift();
+
+  // EqualsArg → "--" ArgName "=" CLIValue ;
+  if (arg.startsWith('--') && arg.includes('=')) {
+    // we're dealing with an EqualsArg, we have a special helper for that
+    const [originalArg, value] = parseEqualsArg(arg);
+    setCLIArg(flags, arg.split('=')[0], normalizeFlagName(originalArg), value);
+    return;
+  }
+
+  // AliasEqualsArg  → "-" AliasName "=" CLIValue ;
+  if (arg.startsWith('-') && arg.includes('=')) {
+    // we're dealing with an EqualsArg, we have a special helper for that
+    const [originalArg, value] = parseEqualsArg(arg);
+    setCLIArg(flags, arg.split('=')[0], originalArg.replace(/^-/, ''), value);
+    return;
   }
 
-  args.forEach((cmdArg) => {
-    let value;
+  // AliasArg → "-" AliasName ( " " CLIValue )? ;
+  if (CLI_FLAG_REGEX.test(arg)) {
+    // this is a short alias, like `-c` for Config
+    setCLIArg(flags, arg, arg.replace(/^-/, ''), parseCLIValue(args));
+    return;
+  }
 
-    if (cmdArg === `--${configCaseName}` || cmdArg === `--${dashCaseName}`) {
-      value = true;
-    } else if (cmdArg === `--no-${dashCaseName}` || cmdArg === `--no${dashToPascalCase(dashCaseName)}`) {
-      value = false;
-    } else if (alias && cmdArg === `-${alias}`) {
-      value = true;
-    }
+  // NegativeDashArg → "--no-" ArgName ;
+  if (arg.startsWith('--no-')) {
+    // this is a `NegativeDashArg` term, so we need to normalize the negative
+    // flag name and then set an appropriate value
+    const normalized = normalizeNegativeFlagName(arg);
+    // TODO should this be a string or a boolean?
+    setCLIArg(flags, arg, normalized, '');
+    return;
+  }
 
-    if (value !== undefined && cmdArg !== undefined) {
-      flags[configCaseName] = value;
-      flags.knownArgs.push(cmdArg);
-    }
-  });
+  // NegativeArg → "--no" ArgName ;
+  if (
+    arg.startsWith('--no') &&
+    !readOnlyArrayHasStringMember(BOOLEAN_CLI_FLAGS, normalizeFlagName(arg)) &&
+    readOnlyArrayHasStringMember(BOOLEAN_CLI_FLAGS, normalizeNegativeFlagName(arg))
+  ) {
+    // possibly dealing with a `NegativeArg` here. There is a little ambiguity
+    // here because we have arguments that already begin with `no` like
+    // `notify`, so we need to test if a normalized form of the raw argument is
+    // a valid and supported boolean flag.
+    const normalized = normalizeNegativeFlagName(arg);
+    setCLIArg(flags, arg, normalized, '');
+    return;
+  }
+
+  // SimpleArg → "--" ArgName ( " " CLIValue )? ;
+  if (arg.startsWith('--')) {
+    setCLIArg(flags, arg, normalizeFlagName(arg), parseCLIValue(args));
+    return;
+  }
+
+  // we shouldn't get here if the input is well formed! so if we do we should
+  // raise an error with the malformed argument
```

Comment:
I have here (as well as below) some points noted where you'd want to raise an error since your parser has encountered something unexpected - however, at present we don't have any error reporting capability in this module, so wanted to raise it for discussion before I added it. It might be nice to not have this module depend on other parts of the codebase and be a little less coupled, so we could `throw` to signal an error. Thoughts?

Reply:
yeah basically something like that, to just let the user that they're trying to use a flag which isn't supported

Reply:
just pushed a change to add error reporting by throwing errors, lmk what you think!